x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalLogicalPlanOptimizer.java

```diff
@@ -162,6 +163,15 @@ else if (plan instanceof Project project) {
                     plan = new Eval(project.source(), project.child(), new ArrayList<>(nullLiteral.values()));
                     plan = new Project(project.source(), plan, newProjections);
                 }
+            } else if (plan instanceof MvExpand) {
+                // We cannot replace the target (NamedExpression) with a Literal
+                // https://github.com/elastic/elasticsearch/issues/109974
+                // Unfortunately we cannot remove the MvExpand right away, or we'll lose the output field (layout problems)
+                // TODO but this could be a follow-up optimization
+                plan = plan.transformExpressionsOnlyUp(
+                    FieldAttribute.class,
+                    f -> stats.exists(f.qualifiedName()) ? f : new Alias(f.source(), f.name(), null, Literal.of(f, null), f.id())
+                );
             }
             // otherwise transform fields in place
             else {
```

Comment:
Ehm, isn't the default case here just asking for trouble? We're replacing field attributes by literals; that's only safe to do further down in the expression tree (if at all). I wonder if we rather shouldn't optimize at all in these cases; or maybe have a version of `transformExpressionsOnlyUp` that never applies to the top-level expression. (Although that would still be unsafe in general, but could be acceptable here.)

Reply:
That's a reasonable concern, probably we should not optimize at all

Reply:
OK, I removed the optimization for now

Reply:
Oh, you mean the final `else`. Yes, that's fragile and could lead to more similar failures when we add new commands.

I didn't change it because it's needed to trigger a few more downstream optimizations, see all the `LocalLogicalPlanOptimizerTests.testMissingField*` tests, and because it's relatively safe with current feature set.

It's worth remembering that it's dangerous though, I'll open an issue so that we can revisit it.

Reply:
https://github.com/elastic/elasticsearch/issues/110150